A36-xds-for-servers.md

```diff
@@ -0,0 +1,401 @@
+is monolithic and we are fine with xDS using internal APIs, there is quite a bit
+more flexibility in design options available for C than Java and Go.
+
+Create an `XdsServerBuilder` that mirrors the `ServerBuilder` API. It may be
+possible to implement it via delegating to a `ServerBuilder` instance, but that
+is an implementation detail.
+
+The `XdsServerBuilder` will need to use a C core API (e.g., a channel arg to
+pass a plugin instance) to enable xDS support in the built `Server`. The
+`XdsServerBuilder` can indirectly create the `XdsClient` and plumb it to filters
+or other code that may need it. Notably, it will _not_ be passed to the
+`ServerCredential`; the `ServerCredential` will be passed the configuration to
+use for each connection, like `ChannelCredential`, so won’t need to use the
+`XdsClient` for watches on the ADS stream. Since many `Server` methods are not
+`virtual`, the `Server` will _not_ be wrapped to add xDS functionality. This
+means the C server will need to be responsible for shutting down the
+`XdsClient`.
+
+Since `ServerBuilder` has few virtual methods, the `XdsServerBuilder` will not
+be directly interchangeable. We believe users will be minimally impacted by
+needing to refer to the precise type, especially with the availability of
+templates. If this becomes a problem more methods could become `virtual`.
+
+
+### Wrapped Languages
+
+This section is a sketch to convey the "feel" of the API. But details may vary.
+
+C will expose an API for wrapped languages to utilize. We do not focus on that
+here and leave that as an implementation detail to be part of a future gRFC.
+
+Each wrapped language will need a custom "xds server" creation API. This will
+vary per-language. We use Python here just as an example of how it may look.
+
+Python could add a `grpc.xds_server(...)` that mirrors `grpc.server(...)`. Since
+ports are added after the server is created, the returned `grpc.Server` object
+may be an xds-aware `Server` to coordinate with the C API when `server.start()`
+is called. However, its implementation should be trivial as it could mostly
+delegate to a normal `server` instance.
+
+It would be possible to manage the lifetime of the XdsClient from the `server`,
+based on `server.stop()`, but given C++ lacks this avenue the C API probably
+will not need this notification.
+
+
+### Java
+
+Create an `XdsServerBuilder` that extends `ServerBuilder` and delegates to a
+"real" builder. The server returned from `builder.build()` would be an xDS-aware
+server delegating to a "real" server instance. The `XdsServerBuilder` would
+install an `AtomicReference<ServerInterceptor>`-backed interceptor in the
+built server and pass the listening port and interceptor reference to the
+xDS-aware server when constructed. Since interceptors cannot be removed once
+installed, the builder may only be used once; `build()` should throw if called a
+second time. To allow the `XdsServerBuilder` to create addition servers if
```

Comment:
s/addition/additional/

But I don't understand this. If subsequent calls to `build()` should throw then how do you propose allowing creation of additional servers with the same builder? Or is that meant to be a future enhancement?

Reply:
Maybe `XdsServer` there would be more clear. I can also drop a hint that this relates to "serving"/"not serving." The point is for the implementation to be able to construct more servers for internal use, not to pass more servers to the user.

Reply:
Okay, got it. The "not serving" to "serving" transition will requires XdsServer to rebuild a new server to be able to serve. Not sure if the internal delegate NettyServerBuilder allows more calls to build.